File: mockgcp/mockaiplatform/metadatastore.go
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// +tool:mockgcp-support
// proto.service: google.cloud.aiplatform.v1beta1.MetadataService
// proto.message: google.cloud.aiplatform.v1beta1.MetadataStore

package mockaiplatform

import (
	"context"
	"fmt"
	"strings"
	"time"

	longrunning "google.golang.org/genproto/googleapis/longrunning"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/emptypb"
	"google.golang.org/protobuf/types/known/timestamppb"

	"github.com/GoogleCloudPlatform/k8s-config-connector/mockgcp/common/projects"
	pb "github.com/GoogleCloudPlatform/k8s-config-connector/mockgcp/generated/mockgcp/cloud/aiplatform/v1beta1"
	"github.com/GoogleCloudPlatform/k8s-config-connector/mockgcp/pkg/storage"
)

type metadataStoreService struct {
	*MockService
	pb.UnimplementedMetadataServiceServer
}

func (s *metadataStoreService) GetMetadataStore(ctx context.Context, req *pb.GetMetadataStoreRequest) (*pb.MetadataStore, error) {
	name, err := s.parseMetadataStoreName(req.Name)
	if err != nil {
		return nil, err
	}

	fqn := name.String()

	obj := &pb.MetadataStore{}
	if err := s.storage.Get(ctx, fqn, obj); err != nil {
		return nil, err
	}
	if obj.State == nil {
		obj.State = &pb.MetadataStore_MetadataStoreState{
			DiskUtilizationBytes: 1,
		}
	}

	return obj, nil
}

func (s *metadataStoreService) CreateMetadataStore(ctx context.Context, req *pb.CreateMetadataStoreRequest) (*longrunning.Operation, error) {
	reqName := req.Parent + "/metadataStores/" + req.MetadataStoreId
	name, err := s.parseMetadataStoreName(reqName)
	if err != nil {
		return nil, err
	}

	fqn := name.String()

	now := time.Now()

	obj := proto.Clone(req.MetadataStore).(*pb.MetadataStore)
	obj.Name = fqn

	obj.CreateTime = timestamppb.New(now)
	obj.UpdateTime = timestamppb.New(now)

	if err := s.storage.Create(ctx, fqn, obj); err != nil {
		return nil, err
	}

	op := &pb.CreateMetadataStoreOperationMetadata{}
	op.GenericMetadata = &pb.GenericOperationMetadata{
		CreateTime: timestamppb.New(now),
		UpdateTime: timestamppb.New(now),
	}
	opPrefix := name.String()
	return s.operations.StartLRO(ctx, opPrefix, op, func() (proto.Message, error) {
		// Many fields are not populated in the LRO result
		result := proto.Clone(obj).(*pb.MetadataStore)
		result.CreateTime = nil
		result.UpdateTime = nil
		return result, nil
	})
}

func (s *metadataStoreService) ListSchemas(ctx context.Context, req *pb.ListMetadataStoresRequest) (*pb.ListMetadataStoresResponse, error) {
	project, err := s.Projects.GetProjectByID(req.Parent)
	if err != nil {
		return nil, err
	}

	findPrefix := fmt.Sprintf("projects/%v/", project.ID)

	var metadataStores []*pb.MetadataStore

	metadataStoreKind := (&pb.MetadataStore{}).ProtoReflect().Descriptor()
	if err := s.storage.List(ctx, metadataStoreKind, storage.ListOptions{}, func(obj proto.Message) error {
		metadataStore := obj.(*pb.MetadataStore)
		if strings.HasPrefix(metadataStore.GetName(), findPrefix) {
			metadataStores = append(metadataStores, metadataStore)
		}
		return nil
	}); err != nil {
		return nil, err
	}

	return &pb.ListMetadataStoresResponse{
		MetadataStores: metadataStores,
	}, nil
}

func (s *metadataStoreService) DeleteMetadataStore(ctx context.Context, req *pb.DeleteMetadataStoreRequest) (*longrunning.Operation, error) {
	name, err := s.parseMetadataStoreName(req.GetName())
	if err != nil {
		return nil, err
	}

	fqn := name.String()
	now := time.Now()

	deleted := &pb.MetadataStore{}
	if err := s.storage.Delete(ctx, fqn, deleted); err != nil {
		return nil, err
	}

	op := &pb.DeleteOperationMetadata{}
	op.GenericMetadata = &pb.GenericOperationMetadata{
		CreateTime: timestamppb.New(now),
		UpdateTime: timestamppb.New(now),
	}
	opPrefix := fmt.Sprintf("projects/%d/locations/%s/metadataStores/%s", name.Project.Number, name.Location, name.MetadataStoreId)
	return s.operations.DoneLRO(ctx, opPrefix, op, &emptypb.Empty{})
}

type MetadataStoreName struct {
	Project         *projects.ProjectData
	Location        string
	MetadataStoreId string
}

func (n *MetadataStoreName) String() string {
	return fmt.Sprintf("projects/%d/locations/%s/metadataStores/%s", n.Project.Number, n.Location, n.MetadataStoreId)
}

// parseMetadataStoreName parses a string into a MetadataStoreName.
// The expected form of input string is projects/<projectID>/locations/<location>/metadataStores/<metadataStoreID>
func (s *MockService) parseMetadataStoreName(name string) (*MetadataStoreName, error) {
	tokens := strings.Split(name, "/")

	if len(tokens) == 6 && tokens[0] == "projects" && tokens[2] == "locations" && tokens[4] == "metadataStores" {
		projectName, err := projects.ParseProjectName(tokens[0] + "/" + tokens[1])
		if err != nil {
			return nil, err
		}
		project, err := s.Projects.GetProject(projectName)
		if err != nil {
			return nil, err
		}

		name := &MetadataStoreName{
			Project:         project,
			Location:        tokens[3],
			MetadataStoreId: tokens[5],
		}

		return name, nil
	} else {
		return nil, status.Errorf(codes.InvalidArgument, "name %q is not valid", name)
	}
}

File: python/services/vertex/alpha/metadata_store.py
# Copyright 2022 Google LLC. All Rights Reserved.
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from connector import channel
from google3.cloud.graphite.mmv2.services.google.vertex import metadata_store_pb2
from google3.cloud.graphite.mmv2.services.google.vertex import metadata_store_pb2_grpc

from typing import List


class MetadataStore(object):
    def __init__(
        self,
        name: str = None,
        create_time: str = None,
        update_time: str = None,
        encryption_spec: dict = None,
        description: str = None,
        state: dict = None,
        project: str = None,
        location: str = None,
        service_account_file: str = "",
    ):

        channel.initialize()
        self.name = name
        self.encryption_spec = encryption_spec
        self.description = description
        self.project = project
        self.location = location
        self.service_account_file = service_account_file

    def apply(self):
        stub = metadata_store_pb2_grpc.VertexAlphaMetadataStoreServiceStub(
            channel.Channel()
        )
        request = metadata_store_pb2.ApplyVertexAlphaMetadataStoreRequest()
        if Primitive.to_proto(self.name):
            request.resource.name = Primitive.to_proto(self.name)

        if MetadataStoreEncryptionSpec.to_proto(self.encryption_spec):
            request.resource.encryption_spec.CopyFrom(
                MetadataStoreEncryptionSpec.to_proto(self.encryption_spec)
            )
        else:
            request.resource.ClearField("encryption_spec")
        if Primitive.to_proto(self.description):
            request.resource.description = Primitive.to_proto(self.description)

        if Primitive.to_proto(self.project):
            request.resource.project = Primitive.to_proto(self.project)

        if Primitive.to_proto(self.location):
            request.resource.location = Primitive.to_proto(self.location)

        request.service_account_file = self.service_account_file

        response = stub.ApplyVertexAlphaMetadataStore(request)
        self.name = Primitive.from_proto(response.name)
        self.create_time = Primitive.from_proto(response.create_time)
        self.update_time = Primitive.from_proto(response.update_time)
        self.encryption_spec = MetadataStoreEncryptionSpec.from_proto(
            response.encryption_spec
        )
        self.description = Primitive.from_proto(response.description)
        self.state = MetadataStoreState.from_proto(response.state)
        self.project = Primitive.from_proto(response.project)
        self.location = Primitive.from_proto(response.location)

    def delete(self):
        stub = metadata_store_pb2_grpc.VertexAlphaMetadataStoreServiceStub(
            channel.Channel()
        )
        request = metadata_store_pb2.DeleteVertexAlphaMetadataStoreRequest()
        request.service_account_file = self.service_account_file
        if Primitive.to_proto(self.name):
            request.resource.name = Primitive.to_proto(self.name)

        if MetadataStoreEncryptionSpec.to_proto(self.encryption_spec):
            request.resource.encryption_spec.CopyFrom(
                MetadataStoreEncryptionSpec.to_proto(self.encryption_spec)
            )
        else:
            request.resource.ClearField("encryption_spec")
        if Primitive.to_proto(self.description):
            request.resource.description = Primitive.to_proto(self.description)

        if Primitive.to_proto(self.project):
            request.resource.project = Primitive.to_proto(self.project)

        if Primitive.to_proto(self.location):
            request.resource.location = Primitive.to_proto(self.location)

        response = stub.DeleteVertexAlphaMetadataStore(request)

    @classmethod
    def list(self, project, location, service_account_file=""):
        stub = metadata_store_pb2_grpc.VertexAlphaMetadataStoreServiceStub(
            channel.Channel()
        )
        request = metadata_store_pb2.ListVertexAlphaMetadataStoreRequest()
        request.service_account_file = service_account_file
        request.Project = project

        request.Location = location

        return stub.ListVertexAlphaMetadataStore(request).items

    def to_proto(self):
        resource = metadata_store_pb2.VertexAlphaMetadataStore()
        if Primitive.to_proto(self.name):
            resource.name = Primitive.to_proto(self.name)
        if MetadataStoreEncryptionSpec.to_proto(self.encryption_spec):
            resource.encryption_spec.CopyFrom(
                MetadataStoreEncryptionSpec.to_proto(self.encryption_spec)
            )
        else:
            resource.ClearField("encryption_spec")
        if Primitive.to_proto(self.description):
            resource.description = Primitive.to_proto(self.description)
        if Primitive.to_proto(self.project):
            resource.project = Primitive.to_proto(self.project)
        if Primitive.to_proto(self.location):
            resource.location = Primitive.to_proto(self.location)
        return resource


class MetadataStoreEncryptionSpec(object):
    def __init__(self, kms_key_name: str = None):
        self.kms_key_name = kms_key_name

    @classmethod
    def to_proto(self, resource):
        if not resource:
            return None

        res = metadata_store_pb2.VertexAlphaMetadataStoreEncryptionSpec()
        if Primitive.to_proto(resource.kms_key_name):
            res.kms_key_name = Primitive.to_proto(resource.kms_key_name)
        return res

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return None

        return MetadataStoreEncryptionSpec(
            kms_key_name=Primitive.from_proto(resource.kms_key_name),
        )


class MetadataStoreEncryptionSpecArray(object):
    @classmethod
    def to_proto(self, resources):
        if not resources:
            return resources
        return [MetadataStoreEncryptionSpec.to_proto(i) for i in resources]

    @classmethod
    def from_proto(self, resources):
        return [MetadataStoreEncryptionSpec.from_proto(i) for i in resources]


class MetadataStoreState(object):
    def __init__(self, disk_utilization_bytes: int = None):
        self.disk_utilization_bytes = disk_utilization_bytes

    @classmethod
    def to_proto(self, resource):
        if not resource:
            return None

        res = metadata_store_pb2.VertexAlphaMetadataStoreState()
        if Primitive.to_proto(resource.disk_utilization_bytes):
            res.disk_utilization_bytes = Primitive.to_proto(
                resource.disk_utilization_bytes
            )
        return res

    @classmethod
    def from_proto(self, resource):
        if not resource:
            return None

        return MetadataStoreState(
            disk_utilization_bytes=Primitive.from_proto(
                resource.disk_utilization_bytes
            ),
        )


class MetadataStoreStateArray(object):
    @classmethod
    def to_proto(self, resources):
        if not resources:
            return resources
        return [MetadataStoreState.to_proto(i) for i in resources]

    @classmethod
    def from_proto(self, resources):
        return [MetadataStoreState.from_proto(i) for i in resources]


class Primitive(object):
    @classmethod
    def to_proto(self, s):
        if not s:
            return ""
        return s

    @classmethod
    def from_proto(self, s):
        return s
